Added models for CourseFacilitator, TeacherProgressReport, and

aoby opened the pull request:
WorkshopOrganizerReport plus associated tests.

Added some extra validations and tests for existing pd models.

For report context, here's the [spec](https://docs.google.com/document/d/1DTu72puX8jAh1LOxGyRjDAc7wL-2P5mBZpm7GmPi15Y).

cc @joshlory.

mehalshah commented on dashboard/app/models/pd/workshop.rb:
```diff
@@ -35,16 +35,39 @@ class Pd::Workshop < ActiveRecord::Base
+    COURSE_COUNSELOR_ADMIN = 'Counselor/Admin'
   ]
 
   STATE_NOT_STARTED = 'Not Started'
```
Might be nicer to define this as

STATES = [
NOT_STARTED,
IN_PROGRESS,
ENDED
] for better readability

aoby replied:
good suggestion. will update.

mehalshah commented on dashboard/app/models/pd/teacher_progress_report.rb:
```diff
@@ -0,0 +1,56 @@
+    teachers =
+      if user.admin?
+        # All teachers in all workshops
+        Pd::Attendance.distinct_teachers
```
This is potentially very large. Does this need to be paginated?

aoby replied:
That's a lower priority work item, for after facilitator summit. We plan to add a date filter to all the reports and probably pagination as well.

mehalshah commented on dashboard/app/models/pd/teacher_progress_report.rb:
```diff
@@ -0,0 +1,56 @@
+          Pd::Attendance.for_workshop(workshop).distinct_teachers
+        end.flatten
+      else
+        []
```
If you get here then something is wrong with the permissions model. You should redirect to forbidden

aoby replied:
good call. will update.

aoby replied:
On second thought, this is only the model. The controller (forthcoming PR) will handle the permissions. I can't redirect at this level, so I'll just raise an error.

mehalshah commented on dashboard/app/models/pd/workshop_organizer_report.rb:
```diff
@@ -0,0 +1,120 @@
+      end
+
+      # 50 * num teachers who have completed > 10 puzzles
+      payment_amount = 50 * num_teachers_qualified
```
Define 50 as a constant somewhere

mehalshah commented on dashboard/app/models/pd/workshop_organizer_report.rb:
```diff
@@ -0,0 +1,120 @@
+      Pd::Workshop::TYPE_PRIVATE
+    ].include?(workshop.workshop_type)
+
+      qualified = teachers.count > 0
```
Same with these constants - move as many of them as possible to constants or even a file that defines payment terms constants

mehalshah commented on dashboard/app/models/pd/workshop_organizer_report.rb:
```diff
@@ -0,0 +1,120 @@
+      payment_amount = (payment_amount + 500 * workshop.facilitators.count) * workshop.sessions.count
+    end
+
+    if workshop.course == Pd::Workshop::COURSE_COUNSELOR_ADMIN &&
```
Make this an elsif

mehalshah commented on dashboard/app/models/pd/workshop_organizer_report.rb:
```diff
@@ -0,0 +1,120 @@
+    }
+  end
+
+  def self.calculate_payment(workshop, teachers, plp)
```
Payment calculation belongs in its own class, and arguably multiple classes.  At the risk of nitpicking, it seems to me that reports should display results more than calculate those results in the first place.

Happy to talk more about it offline

aoby commented:
I updated the PR with the following changes:
1. Moved workshop organizer payment model into its own class. I clarified
   the various payment part names with PM and now expose each one so it's
   easier to follow, test, and reason about.
2. Renamed some of the strings for workshop course and subject on
   request from PM.
3. Incorporated earlier PR feedback.

mehalshah commented on dashboard/app/models/pd/session.rb:
```diff
@@ -18,6 +18,24 @@ class Pd::Session < ActiveRecord::Base
+  validate :starts_and_ends_on_the_same_day
+  validate :starts_before_ends
+
+  def starts_and_ends_on_the_same_day
```
While you are doing this you should have some testing for this

aoby replied:
tests added

mehalshah commented:
This looks fine but the payment calculation part is at the upper bound of desirable complexity. I think this is good to check in and revisit the calculators later

joshlory commented on dashboard/app/models/pd/attendance.rb:
```diff
@@ -19,7 +19,19 @@ class Pd::Attendance < ActiveRecord::Base
-    joins(:workshop).where(teacher: teacher, pd_workshops: {id: workshop.id})
+  def self.for_teacher(teacher)
+    self.joins(:workshop).where(teacher_id: teacher.id)
+  end
```
Nit: prefer [scopes](http://api.rubyonrails.org/classes/ActiveRecord/Scoping/Named/ClassMethods.html#method-i-scope) over class methods here.  (Scopes are just syntactic sugar for the class method.)
